File: scripts/me.rb
#!/usr/bin/env ruby
# output sive.rs static site
require 'cgi'
require 'pg'
DB = PG::Connection.new(dbname: 'sivers', user: 'sivers')

ODIR = '/var/www/html/sive.rs/' 
FDIR = '/home/derek/code/b/public/sive.rs/'
%x(rsync -a #{FDIR}* #{ODIR})

# blog body can have this line:
# <!-- CODE: filename.here -->
# replace with HTML of that code from code/
def code_replace(body)
  return body unless body.include?('<!-- CODE: ')
  r = %r{<!-- CODE: (\S+) -->}
  codedir = ODIR + 'code/'
  newbody = ''
  body.split("\n").each do |line|
    m = r.match(line.strip)
    unless m
      newbody << (line + "\n")
    else
      filename = m[1]
      raise "code missing: #{filename}" unless File.exist?(codedir + filename)
      # first line of code is credit pointing to article URL
      code = File.readlines(codedir + filename)[1..-1].join('').strip
      newbody << '<div class="code">'
      newbody << "\n<pre><code>"
      newbody << CGI.escapeHTML(code)
      newbody << "\n</code></pre>"
      newbody << '<small><a href="/code/%s">download code</a></small>' % filename;
      newbody << "\n</div>\n"
    end
  end
  newbody
end

def q2o(from, uri)
  html = DB.exec("select body from me.#{from}")[0]['body']
  canon = '<link rel="canonical" href="https://sive.rs/%s">' % (uri == 'index.html' ? '' : uri)
  html.gsub!(/^<title>/, canon + "\n<title>")
  File.write(ODIR + uri, code_replace(html))
end

DB.exec("select uri from me.article_uris()").each do |o|
  q2o("article('%s')" % o['uri'], o['uri'])
end

q2o("articles()", "blog")

# excluding metabook pages since they are still static
DB.exec("select uri from topics where uri not in (select uri from metabooks)").each do |o|
  q2o("topic_page('%s')" % o['uri'], o['uri'])
end

%x(mkdir -p #{ODIR}/book)
DB.exec("select uri from me.book_uris()").each do |o|
  q2o("book('%s')" % o['uri'], "book/%s" % o['uri'])
end

q2o("books()", "book/index.html")

q2o("home()", "index.html")

DB.exec("select uri from me.interview_uris()").each do |o|
  q2o("interview('%s')" % o['uri'], o['uri'])
end

q2o("interviews()", "i")

%x(mkdir -p #{ODIR}/met)
q2o("met()", "met/index.html")

DB.exec("select id from me.met1_ids()").each do |o|
  q2o("met1(%d)" % o['id'], "met/%d" % o['id'])
end

DB.exec("select id from me.metat_ids()").each do |o|
  q2o("metat(%d)" % o['id'], "met/at-%d" % o['id'])
end

DB.exec("select uri, pagetitle from me.pages()").each do |o|
  q2o("page('%s', '%s')" % [o['uri'], o['pagetitle']], o['uri'])
end

DB.exec("select uri from me.presentation_uris()").each do |o|
  q2o("presentation('%s')" % o['uri'], o['uri'])
end

q2o("presentations()", "presentations")

q2o("refs()", "ref")

q2o("tweets()", "d")

q2o("sitemap()", "sitemap.xml")

File: scripts/me.go
package main

// write sive.rs static site: go run scripts/me.go
import (
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sive.rs/sivers/internal/xx"
)

// thank you, blunt checker
func check(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	check(xx.InitDB(true))
	defer xx.DB.Close()

	const out = "/var/www/html/sive.rs/"
	const files = "/home/derek/code/b/public/sive.rs/"
	cmd := exec.Command("rsync", "-a", files, out)
	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
	check(cmd.Run())

	// hack to add link rel canonical to header at the moment of file-writing:
	// since writer (below) knows uri, find <title> and replace it with <link...>\n<title> 
	title := regexp.MustCompile(`(?m)^<title>`)

	// clever clever: getting the output uri along with the query of contents (uri, html)
	for _, query := range []string{
		`select uri, me.article(uri) from me.article_uris()`,
		`select 'blog', me.articles()`,
		`select uri, me.topic_page(uri) from topics where uri not in (select uri from metabooks)`,
		`select 'book/' || uri, me.book(uri) from me.book_uris()`,
		`select 'book/index.html', me.books()`,
		`select 'index.html', me.home()`,
		`select uri, me.interview(uri) from me.interview_uris()`,
		`select 'i', me.interviews()`,
		`select 'met/index.html', me.met()`,
		`select 'met/' || id, me.met1(id) from me.met1_ids()`,
		`select 'met/at-' || id, me.metat(id) from me.metat_ids()`,
		`select uri, me.page(uri, pagetitle) from me.pages()`,
		`select uri, me.presentation(uri) from me.presentation_uris()`,
		`select 'presentations', me.presentations()`,
		`select 'ref', me.refs()`,
		`select 'd', me.tweets()`,
		`select 'sitemap.xml', me.sitemap()`,
	} {
		rows, err := xx.DB.Query(query)
		check(err)
		for rows.Next() {
			var uri, html string
			check(rows.Scan(&uri, &html))
			// because sive.rs/index.html and sive.rs/book/index.html are not the canonical URL, erase that bit:
			url := strings.ReplaceAll(uri, "/index.html", "")
			html = title.ReplaceAllLiteralString(html, `<link rel="canonical" href="https://sive.rs/`+url+`">`+"\n<title>")
			check(os.MkdirAll(filepath.Dir(out+uri), 0755))
			check(os.WriteFile(out+uri, []byte(html), 0644))
		}
		check(rows.Err())
		check(rows.Close())
	}
}
